File: js/t1/tests/model.properties.test.js
import { describe, test } from 'node:test';
import assert from 'node:assert/strict';
import fc from 'fast-check';
import { T } from '../model.js';
import * as TI from './testInstance.js';

const eng = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
              TI.PW, TI.FY, TI.FX);

// Event generator: random sequence of {kind, args} over the four non-trivial events.
const pieceArb = fc.constantFrom(...TI.Piece);
const eventArb = fc.oneof(
  fc.record({ kind: fc.constant('move'), dy: fc.constantFrom(-1, 0), dx: fc.constantFrom(-1, 0, 1) }),
  fc.record({ kind: fc.constant('rotate'), cw: fc.boolean() }),
  fc.record({ kind: fc.constant('fix'), pNew: pieceArb }),
  fc.record({ kind: fc.constant('fall'), pNew: pieceArb }),
);

function applyEvent(m, e) {
  switch (e.kind) {
    case 'move': return m.movePiece(e.dy, e.dx);
    case 'rotate': return m.rotatePiece(e.cw);
    case 'fix': return m.fixPiece(e.pNew);
    case 'fall': return m.fallStep(e.pNew);
  }
}

describe('T1 model.js properties', () => {
  test('invariants hold after every step of every random trace', () => {
    fc.assert(fc.property(pieceArb, fc.array(eventArb, { maxLength: 200 }), (p0, events) => {
      const m = new eng.Machine(p0);
      let sawGameover = false;
      for (const e of events) {
        assert.doesNotThrow(() => applyEvent(m, e));               // D9 guard-before-index
        assert.strictEqual(eng.typeOK(m), true);                    // typeOK(snapshot)
        assert.strictEqual(m.mg.length, TI.InitialMainGrid.length); // mg stays HM x WM
        assert.strictEqual(m.mg[0].length, TI.InitialMainGrid[0].length);
        assert.doesNotThrow(() => eng.checkInvariants(m, 'property'));

        // integer fields within MAX_SAFE_INTEGER (D10)
        for (const f of [m.py, m.px, m.pr, m.clearedLines])
          assert.ok(Math.abs(f) <= Number.MAX_SAFE_INTEGER);

        // gameover monotonicity
        if (sawGameover) assert.strictEqual(m.gameover, true);
        if (m.gameover) sawGameover = true;
      }
    }), { numRuns: 200 });
  });

  test('clearFullLines is idempotent and dimension-preserving', () => {
    fc.assert(fc.property(
      fc.array(fc.array(fc.boolean(), { minLength: 3, maxLength: 3 }), { minLength: 1, maxLength: 8 }),
      (grid) => {
        const once = eng.clearFullLines(grid);
        const twice = eng.clearFullLines(once);
        assert.deepStrictEqual(once, twice);
        assert.strictEqual(once.length, grid.length);
        assert.strictEqual(once[0].length, grid[0].length);
      }));
  });

  test('union dimensions match g1 dimensions', () => {
    fc.assert(fc.property(
      fc.array(fc.array(fc.boolean(), { minLength: 2, maxLength: 2 }), { minLength: 2, maxLength: 2 }),
      fc.array(fc.array(fc.boolean(), { minLength: 2, maxLength: 2 }), { minLength: 2, maxLength: 2 }),
      fc.integer({ min: -2, max: 2 }), fc.integer({ min: -2, max: 2 }),
      (g1, g2, y1, x1) => {
        const u = eng.union(g1, y1, x1, g2, 0, 0);
        assert.strictEqual(u.length, g1.length);
        assert.strictEqual(u[0].length, g1[0].length);
      }));
  });
});

File: rs/t1/tests/model_properties_test.rs
//! spec: tests/model_properties_test.rs — §9.2, `implementation.md`.

#[path = "test_instance.rs"]
mod fixture;

use fixture::{Piece as TPiece, TestInstance};
use proptest::prelude::*;
use t1::model::{self, Machine, Params, PieceOrExtra};

const HM: i64 = 6; // TestInstance's board dimensions (tests/test_instance.rs)
const WM: i64 = 5;
const B: i64 = 8; // max(HM,WM) + PW - 1 = max(6,5) + 3 - 1 (D10)

fn check_common_invariants(m: &Machine<TestInstance>) {
    // type_ok after every step
    assert!(model::type_ok(m), "type_ok violated: {:?}", m.mg);
    // mg stays HM×WM
    assert_eq!(m.mg.len() as i64, HM);
    for row in &m.mg {
        assert_eq!(row.len() as i64, WM);
    }
    // every integer field within ±B (D10)
    assert!(m.py.abs() <= B, "py={} out of ±{B}", m.py);
    assert!(m.px.abs() <= B, "px={} out of ±{B}", m.px);
    // check_invariants: Correct's five conjuncts, re-checked every step
    model::check_invariants(m);
}

/// One step of a random trace, dispatched by a small integer command plus a
/// piece choice (only consumed by `Fix`/`Fall`).
fn apply_command(m: &mut Machine<TestInstance>, cmd: u8, piece: TPiece) {
    match cmd % 7 {
        0 => {
            m.move_piece(0, -1);
        }
        1 => {
            m.move_piece(0, 1);
        }
        2 => {
            m.move_piece(-1, 0);
        }
        3 => {
            m.rotate_piece(true);
        }
        4 => {
            m.rotate_piece(false);
        }
        5 => {
            m.fix_piece(piece);
        }
        _ => {
            m.fall_step(piece);
        }
    }
}

fn piece_from_index(i: u8) -> TPiece {
    TestInstance::piece_all()[(i % 2) as usize]
}

proptest! {
    #![proptest_config(ProptestConfig { cases: 256, .. ProptestConfig::default() })]

    #[test]
    fn invariants_hold_over_random_traces(
        start_piece_idx in 0u8..2,
        trace in prop::collection::vec((0u8..7, 0u8..2), 1..300),
    ) {
        let mut m = Machine::<TestInstance>::new(piece_from_index(start_piece_idx), fixture::piece_source_stub());
        check_common_invariants(&m);
        let mut prev_gameover = m.gameover;

        for (cmd, piece_idx) in trace {
            apply_command(&mut m, cmd, piece_from_index(piece_idx));
            check_common_invariants(&m);
            // gameover monotonicity: never flips back to false once true
            prop_assert!(!(prev_gameover && !m.gameover), "gameover flipped false after being true");
            prev_gameover = m.gameover;
        }
    }
}

// ── fix_piece's swap-partition: buffer reuse, not reallocation ──────────
//
// D13: line-clearing is fused in place, not a fresh allocation. This checks
// that every surviving row keeps its original buffer (by address), just
// possibly reordered or reset in place.

proptest! {
    #![proptest_config(ProptestConfig { cases: 128, .. ProptestConfig::default() })]

    #[test]
    fn fix_piece_reuses_row_buffers(
        // which of the 5 rows *other than* row 0 (the piece's resting row)
        // to pre-fill as full lines
        full_rows in prop::collection::hash_set(1usize..6, 0..5),
    ) {
        let mut m = Machine::<TestInstance>::new(TPiece::Bar, fixture::piece_source_stub());
        m.py = -1; // rests on row 0 (see model_unit_test.rs's bar_at_rest)
        m.px = 0;
        prop_assert!(!model::can_move_piece(-1, 0, &m));

        for &y in &full_rows {
            m.mg[y] = vec![Some(PieceOrExtra::Piece(TPiece::Corner)); WM as usize];
        }

        let before: std::collections::BTreeSet<usize> =
            m.mg.iter().map(|row| row.as_ptr() as usize).collect();
        prop_assert_eq!(before.len(), HM as usize, "rows should have distinct buffers to begin with");

        prop_assert!(m.fix_piece(TPiece::Corner));
        prop_assert_eq!(m.cleared_lines as usize, full_rows.len());

        let after: std::collections::BTreeSet<usize> =
            m.mg.iter().map(|row| row.as_ptr() as usize).collect();
        prop_assert_eq!(before, after, "row buffers should be the same set, just reordered/cleared in place");
    }
}
